resources/prosody-plugins/mod_visitors_component.lua

```diff
@@ -0,0 +1,190 @@
+local function stanza_handler(event)
+    local origin, stanza = event.origin, event.stanza;
+
+    if stanza.name ~= 'iq' then
+        return;
+    end
+
+    if stanza.attr.type == 'result' and sent_iq_cache:get(stanza.attr.id) then
+        sent_iq_cache:set(stanza.attr.id, nil);
+        return true;
+    end
+
+    if stanza.attr.type ~= 'set' then
+        return; -- We do not want to reply to these, so leave.
+    end
+
+    local visitors_iq = event.stanza:get_child('visitors', 'jitsi:visitors');
+    if not visitors_iq then
+        return;
+    end
+
+    if origin.type ~= 's2sin' then
+        module:log('warn', 'not from s2s session, ignore! %s', stanza);
+        return true;
+    end
+
+    local room_jid = visitors_iq.attr.room;
+    local room = get_room_from_jid(room_jid_match_rewrite(room_jid));
+
+    if not room then
+        log('warn', 'No room found %s', room_jid);
+        return;
+    end
+
+    local request_promotion = visitors_iq:get_child('request-promotion');
+    if not request_promotion then
+        return;
+    end
+
+    -- respond with successful receiving the iq
+    origin.send(st.iq({
+        type = "result";
+        from = stanza.attr.to;
+        to = stanza.attr.from;
+        id = stanza.attr.id
+    }));
+
+    -- TODO send iq to moderators (what about name? -> it will be coming from the token) or auto allow
+    if not visitors_promotion_map[room.jid] and auto_allow_promotion then
+        -- visitors is enabled
+        visitors_promotion_map[room.jid] = {};
+    end
+
+    -- if visitors is enabled for the room
+    if visitors_promotion_map[room.jid] then
+        if auto_allow_promotion then
+            --  we are in auto-allow mode, let's reply with accept
+            -- we store where the request is coming from so we can send back the response
+            local username = new_id():lower();
+            visitors_promotion_map[room.jid][username] = {
+                from = stanza.attr.from;
+                jid = request_promotion.attr.jid;
+                room = visitors_iq.attr.room;
+            };
+
+            local req_from = visitors_promotion_map[room.jid][username].from;
+            local req_room = visitors_promotion_map[room.jid][username].room;
+            local req_jid = visitors_promotion_map[room.jid][username].jid;
+            local focus_occupant = get_focus_occupant(room);
+            local focus_jid = focus_occupant and focus_occupant.bare_jid or nil;
+
+            local iq_id = new_id();
+            sent_iq_cache:set(iq_id, socket.gettime());
+
+            module:send(st.iq({
+                type='set',
+                to = req_from,
+                from = module.host,
+                id = iq_id })
+                          :tag('visitors', {
+                xmlns='jitsi:visitors',
+                room = string.gsub(req_room, muc_domain_base, req_from),
+                focusjid = focus_jid })
+                          :tag('response-promotion', {
```

Comment:
nit: Suggestion for names: `request-promotion` and `accept-promotion`, or `promotion-request` and `promotion-response`